File: BinarySearch.cpp
#include <chrono>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

using namespace std;
using namespace std::chrono;

int binarySearch(const vector<string>& data, int low, int high, const string& target) {
    while (low <= high) {
        int mid = low + (high - low) / 2;
        
        int comparison = target.compare(data[mid]);
        if (comparison == 0) {
            return mid;
        }
        
        if (comparison < 0) {
            high = mid - 1;
        } else {
            low = mid + 1;
        }
    }
    return -1;
}

double measureSearchTime(const vector<string>& data, const string& target, int n, int repetitions) {
    double totalTime = 0.0;
    volatile int dummy = 0;
    
    for (int rep = 0; rep < repetitions; rep++) {
        auto start = high_resolution_clock::now();
        
        volatile int result = 0;
        for (int i = 0; i < n; i++) {
            result += binarySearch(data, 0, data.size() - 1, target);
        }

        if (result == -123456789) std::cout << "";
        
        auto stop = high_resolution_clock::now();
        auto duration = duration_cast<microseconds>(stop - start);
        totalTime += duration.count() / 1000.0;

        if (dummy == -123456789) cout << "";
    }
    
    return totalTime / repetitions;
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        cout << "Usage: " << argv[0] << " <input_csv_file>" << endl;
        return 1;
    }
    
    string inputFileName = argv[1];
    ifstream MyFile(inputFileName);
    
    if (!MyFile.is_open()) {
        cout << "Error: Could not open file " << inputFileName << endl;
        return 1;
    }
    
    string line, word;
    vector<string> data;
    
    while (getline(MyFile, line)) {
        stringstream ss(line);
        getline(ss, word, ',');
        data.push_back(word);
    }
    MyFile.close();
    
    int n = data.size();
    const int repetitions = 1;
    
    string bestCase = data[n / 2];
    double bestTime = measureSearchTime(data, bestCase, n, repetitions);
    
    // Calculate average case using multiple positions
    vector<int> positions = {n/8, n/6, n/4, n/3, (2*n)/5, (3*n)/5, (2*n)/3, (3*n)/4};
    double totalAvgTime = 0.0;
    
    for (int pos : positions) {
        string avgCase = data[pos];
        totalAvgTime += measureSearchTime(data, avgCase, n, repetitions);
    }
    double avgTime = totalAvgTime / positions.size();
    
    string worstCase = data[n - 1];
    double worstTime = measureSearchTime(data, worstCase, n, repetitions);
    
    string outputFileName = "binary_search_" + to_string(n) + ".txt";
    ofstream outFile(outputFileName);
    outFile << "Binary Search Performance Analysis\n\n";
    
    outFile << "Number of searches performed for each case: " << n << "\n";
    outFile << "Number of measurement repetitions: " << repetitions << "\n\n";
    outFile << fixed << setprecision(9);
    outFile << "Best case (middle element) average time: " << bestTime << " ms\n";
    outFile << "Average case (average of elements at 1/8, 1/6, 1/4, 1/3, 2/5, 3/5, 2/3, and 3/4 positions) average time: "
            << avgTime << " ms\n";
    outFile << "Worst case (last element) average time: " << worstTime << " ms\n";
    outFile.close();
    
    cout << "Results have been written to " << outputFileName << endl;
    
    return 0;
}

File: BinarySearch.java
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileWriter;
import java.io.IOException;
import java.util.Scanner;
import java.util.ArrayList;

public class BinarySearch {

    public static int binarySearch(ArrayList<String[]> array, String target) {
        int left = 0;
        int right = array.size() - 1;

        while (left <= right) {
            int mid = left + (right - left) / 2;

            if (array.get(mid)[0].equals(target)) {
                return mid;
            } else if (array.get(mid)[0].compareTo(target) < 0) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        return -1;
    }

    public static long measureSearchTime(ArrayList<String[]> array, String target, int iterations) {
        long totalTime = 0;
        
        long startTime = System.nanoTime();
        for (int i = 0; i < iterations; i++) {
            binarySearch(array, target);
        }
        long endTime = System.nanoTime();
        totalTime += (endTime - startTime);
    
        return totalTime;
    }

    public static void main(String[] args) {
        if (args.length < 1) {
            System.out.println("Usage: java BinarySearch <input_csv_file>");
            return;
        }
        
        String inputFileName = args[0];
        ArrayList<String[]> dataList = new ArrayList<>();

        try {
            File myObj = new File(inputFileName);
            Scanner myReader = new Scanner(myObj);
            while (myReader.hasNextLine()) {
                String[] dataArray = myReader.nextLine().split(",");
                dataList.add(dataArray);
            }
            myReader.close();
        } catch (FileNotFoundException e) {
            System.out.println("An error occurred: File not found - " + inputFileName);
            e.printStackTrace();
            return;
        }

        int n = dataList.size();

        for (int i = 0; i < 100000; i++) {
            binarySearch(dataList, dataList.get(i % n)[0]);
        }
        
        // Calculate average case using multiple elements
        int[] positions = {n/8, n/6, n/4, n/3, (2*n)/5, (3*n)/5, (2*n)/3, (3*n)/4};
        long totalAverageCaseTime = 0;
        for (int pos : positions) {
            String element = dataList.get(pos)[0];
            totalAverageCaseTime += measureSearchTime(dataList, element, n);
        }
        long averageCaseTime = totalAverageCaseTime / positions.length;
        
        int last = n - 1;
        String lastElement = dataList.get(last)[0];
        long worstCaseTime = measureSearchTime(dataList, lastElement, n);
        
        int mid = n / 2;
        String middleElement = dataList.get(mid)[0];
        long bestCaseTime = measureSearchTime(dataList, middleElement, n);

        StringBuilder output = new StringBuilder();
        output.append("Binary Search Performance Analysis\n\n");
        output.append(String.format("Number of searches performed for each case: %d\n\n", n));
        output.append(String.format("Best case (middle element) average time: %.6f milliseconds\n", bestCaseTime / 1_000_000.0));
        output.append(String.format("Average case (average of elements at 1/8, 1/6, 1/4, 1/3, 2/5, 3/5, 2/3, and 3/4 positions) average time: %.6f milliseconds\n",
                averageCaseTime / 1_000_000.0));
        output.append(String.format("Worst case (last element) average time: %.6f milliseconds\n", worstCaseTime / 1_000_000.0));

        String outputFileName = "binary_search_" + n + ".txt";
        try {
            FileWriter writer = new FileWriter(outputFileName);
            writer.write(output.toString());
            writer.close();
            System.out.println("Results have been written to " + outputFileName);
        } catch (IOException e) {
            System.out.println("An error occurred while writing to the file.");
            e.printStackTrace();
        }
    }
}